### DISCUSSION TITLE
useReducer

### DISCUSSION THREAD

**Discussion by honaseung** [honaseung/react-deep-dive#26] [Chapter3]
We talked about `useContext`. We can get 'context' from every where under `Context.Provider` if we call it. But is it only thing we can do?  Can't we manipulate it? Of course we do. Sharing between far components is the key point of using `useContext`. So we can manipulate from a component and affect it to the others. It's not that differenct with `state`. But it isn't in a component. `useContext` is not a sort of magic. There is no such super dramatic optimization. You can consider it as `states`. Anyway, You can manipulate it. And manipulating is `useReducer`.

Like I told you, We already familiar with `redux` not `useContext` or `useReducer`. So we are going to look over.
```
const initialState = { count: 0};
const reducer = (state, action) =>{
  switch (action.type) {
    case ...
    .
    .
    .
  }
}
const init = () => {
  return { count: 0 };
}

const [state, dispatcher] = useReducer(reducer, initialState, init);
```
First argument `reducer` is a sort of manipulater. You can give a 'type' as a 'signal'. Then `reducer` manipulate as your order. You can only manipulate by this way. You can't manipulate it as you want.
The second argument `initialState` is an initial value of `state` like the word's meaning.
Thrid one is an optional function argument. If you don't need it, You don't have to use it. When you use it. It's used for `initializing`.  You can give anonymous function if you want, But if you save it by variable, You can also use it from inside of reducer.

I guess it's all I can tell you about `useReducer`. I think it's enough to know.

**Comment by honaseung** [comment]
우리는 `useContext` 에 대해 이야기 해봤습니다. 우리는 `Context.Provider` 아래라면 어디서든 부르기만 하면 'context' 를 얻을 수 있습니다. 하지만 우리는 단지 부를수만 있을까요? 당연하게도 조작할 수도 있습니다. `useContext` 를 사용하는 키 포인트는 멀리 떨어진 두 컴포넌트간의 공유니깐요. 그래서 우리는 하나의 컴포넌트에서 이를 조작하고 다른 컴포넌트들에게 전달할 수 있습니다. `state` 와 크게 다르지 않아요. 단지 컴포넌트 내부에 존재 않을 뿐이죠. `useContext` 는 마법이 아닙니다. 극적인 최적화를 기대해선 안됩니다. 당신은 이것을 `states` 처럼 생각해야 합니다. 어쨋든 당신은 `useReducer` 를 이용하여 이것을 조작할 수 있습니다.

이미 말씀드렸다시피 우리는 `redux` 랑 친하지 `useContext` 나 `useReducer` 와 친하진 않습니다. 그러니 그냥 슥 훑어만 봅시다.
```
const initialState = { count: 0};
const reducer = (state, action) =>{
  switch (action.type) {
    case ...
    .
    .
    .
  }
}
const init = () => {
  return { count: 0 };
}

const [state, dispatcher] = useReducer(reducer, initialState, init);
```
첫번째 인자인 `reducer` 는 조작자입니다. 당신은 '타입'을 '시그널'로써 줄 수 있고 `reducer` 는 당신의 명령대로 동작합니다. 당신은 오직 이런 방법으로만 조작할 수 있습니다. 당신이 원하는대로 조작할 순 없어요.
두번째 인자인 `initialState` 는 문자 그대로 `state` 의 초기값입니다.
세번째 것은 필수는 아닙니다. 필요없으면 안 써도 됩니다. 당신이 이것을 쓰면 이것은 `initailizing` 에 사용됩니다. 당신은 원한다면 익명 함수를 넘겨줄 수도 있지만 함수를 변수에 담아서 사용한다면 `reducer` 내부에서도 호출할 수 있도록 만들 수 있습니다.

**Reply by chaminju927** [reply]
잘 이해가 안가는 부분이 있습니다 ! 그럼 useReducer의 세번째 인자인 init은 초기값을 생성하는 함수를 의미하는 건가요?

**Reply by honaseung** [reply]
조금 비슷합니다!! 그냥 바로 초기 값을 생성할 수도 있고 초기 값을 계산하여 생성하거나 등등 기타 다른 작업들을 할 수 있습니다.
